fix: yield form in fastui_form to keep file handler open

gradient-ascent-ai-lab opened the pull request:
I ran into the same issue as  #146, where in all cases the file handler of an UploadFile is already closed when it becomes available in the endpoint function. I found where the closing of a context manager leads to the closing of the filehandler, and found that yielding inside the `request.form()` context instead of returning after it fixes it. I'm not sure if this is the ideal solution (probably not as CI is breaking on the typechecker) but at least it's pretty minimal and fixes the issue, so I thought it useful enough to share.

gradient-ascent-ai-lab committed:
fix: yield form in fastui_form to keep file handler open

samuelcolvin reviewed (commented):
Thanks so much. Please fix tests and and some more tests that would fail without this.

samuelcolvin commented:
Well the tests having odd syntax isn't the end of the world as long as real world cases are handled correctly and elegantly.

I think the solution might be as simple as adding `@asynccontextmanager` to `run_fastui_form`, then using `async with` in the tests instead of just `await`. To get tests to fail correctly, I guess you'll need to update `FakeRequest` to to someone close any files after they're returned.

gradient-ascent-ai-lab committed:
add asynccontextmanager to run_fastui_form

gradient-ascent-ai-lab committed:
update tests

gradient-ascent-ai-lab commented:
Thanks for the guidance! 
I've added the `@asynccontextmanager`, updated the syntax of the tests to use `async with`, updated the `FakeRequest` to close files after yielding, added `assert not {filehandler}.closed` wherever UploadFile is used, and all tests now pass. 
I've rebased on main, if there's anything else this PR needs let me know.

abuchnick-aiola commented:
@samuelcolvin Any progress on this please ?
Currently I'm using the monkeypatch @gradient-ascent-ai-lab suggested [here](https://github.com/pydantic/FastUI/issues/146), but prefer avoiding this of course.

matthewbelcher commented:
I've run into this issue myself. Is this PR going to make it to master or is there another better fix?

rgimen3z commented:
@sydney-runkle / @samuelcolvin, seems like this PR is ready to be merged and would be 
very useful for many users. Mind taking another look when you have a chance? Thanks
